library/init/meta/widget/basic.lean

```diff
@@ -117,19 +171,41 @@ inductive mouse_event_kind
 | on_mouse_enter
 | on_mouse_leave
 
+@[derive decidable_eq]
+inductive mouse_capture_state
+| outside
+| immediate
+| child
+
 meta mutual inductive component, html, attr
 
 with component : Type → Type → Type
-| mk {Props : Type}
-     {Action : Type}
-     (InnerAction : Type)
-     (State : Type)
-     (init : Props → option State → State)
-     (update : Props → State → InnerAction → State × option Action)
-     (view : Props → State → list (html InnerAction))
-     /- If this returns true, then the component will not call 'view' again. -/
-     (props_eq : Props → Props → bool)
+| pure
+     {Props Action : Type}
+     (view : Props → list (html Action))
      : component Props Action
+| filter_map_action
```

Comment:
My knowledge of React is extremely limited.  These hooks here look more like higher-order components than React hooks.  Am I getting this roughly right?

It looks like most of these hooks could be implemented using a combination of `component.pure` and `component.with_state`.  I assume these are optimizations?

Reply:
Yeah so `with_state` and `with_mouse` would be the same spirit react hooks. But filter_map_action and map_props are better thought of as wrappers for the component. I think these could be implemented not as explicit constructors but at the end of the day I found this was simpler

Reply:
The main point of this change is because it enables adding `with_task` and other gadgets later without having to break the API and to separate concerns a little in the implementation.